### ISSUE TITLE
Update setting data pointers for Cython 3

### ISSUE THREAD

**Issue by jreback** [pandas-dev/pandas#34014]
Need to update the following locations

- [x] _libs/window/aggregations.pyx: `bufarr.data`
- [ ] _libs/reduction.pyx: `chunk.data`

---

https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=34888&view=logs&j=33ccdd52-c922-5ef2-8209-78215e36d994&t=046ffbec-62c2-54e3-e88a-7745f4292bb6&l=457

just started failing.

cc @pandas-dev/pandas-core @pandas-dev/pandas-triage 

if anyone has insights

**Comment by alimcmaster1** [comment]
Cython alpha was released today: https://pypi.org/project/Cython/#history

Locally I get below error when build extensions

...
        buf = <float64_t *>arr.data
        bufarr = np.empty(win, dtype=float)
        oldbuf = <float64_t *>bufarr.data
        for i in range((win - offset), (N - offset)):
            buf = buf + 1
            bufarr.data = <char *>buf
                 ^
------------------------------------------------------------

pandas/_libs/window/aggregations.pyx:1385:18: Assignment to a read-only property
```

**Comment by WillAyd** [comment]
My guess is this is the result of https://github.com/cython/cython/pull/3571

**Comment by WillAyd** [comment]
So IIUC assigning the data pointer has been deprecated since 2013. Quick fix might be to get an upstream revert of the change, but in any case we probably need to fix this.

Not sure there is a true replacement in the more modern API that allows for assigning the data pointer at least according to the following mailing list:

https://mail.python.org/pipermail/numpy-discussion/2013-November/068172.html

ref https://github.com/cython/cython/issues/2498 and ping @rgommers as well

**Comment by scoder** [comment]
That looks like the one evil hack that smells a bit when you implement it but otherwise works perfectly, and that drops on your feet as soon as you've forgotten about it. You already CC-ed NumPy, they are the right people to ask here. CC-ing @mattip as well, who came up with the property wrapper for these fields for Cython.

I'm happy that you didn't ask for an upstream revert of the change. :)

**Comment by bashtage** [comment]
https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/reduction.pyx#L105

is also very fishy.  It also seems to be at the root of a very strange behavior when group by apply is used in statsmodels/statsmodels#6603 .  In this example, the wrong data was ending up in in the OLS regression because it looks like it has the wrong buffer somehow.  This seems to be not completely safe in the case of a deferred operation.

**Comment by mattip** [comment]
Maybe rewrite that to use `bufarr = PyArray_SimpleNewFromData(1, &win, NPY_FLOAT32, buf)` instead of `bufarr.data = ...`? As for the line pointed to by @bashtage, that also needs a rewrite to be safe.

**Comment by mattip** [comment]
The `Reducer` class was added in [a commit](https://github.com/pandas-dev/pandas/commit/74f5d6d52f47ce86a607769df68bdf9c68b7292a) in response to gh-309  in 2011. Perhaps it could be rewritten to use [nditer](https://numpy.org/devdocs/reference/generated/numpy.nditer.html#numpy.nditer) from python  or the [Array Iterator](https://numpy.org/devdocs/reference/c-api/iterator.html) from C (if the former gives performance problems).

**Comment by mattip** [comment]
Of course you could just write `PyArray_Bytes(chunk) = arr.data`, although that would not solve any possible bugs or race conditions due to reusing `chunk` inappropriately

**Comment by jreback** [comment]
ok I think we need to fix the cython version again, @alimcmaster1 would you mind doing a PR for this.

We should certainly address this older code paths. IIRC @jbrockmendel was almost able to remove this entire module, but has some perf issues.

**Comment by TomAugspurger** [comment]
If we don't fix these for 1.1.0 (or 1.0.4), we might consider putting an upper bound on Cython in our pyproject.toml: https://github.com/pandas-dev/pandas/blob/master/pyproject.toml. I think that's the only situation where this would affect users. People installing pandas from source after Cython 3.0 is released.

**Comment by jbrockmendel** [comment]
> was almost able to remove this entire module, but has some perf issues.

Yah, we need to decide how big a perf penalty we're willing to accept in exchange for losing the maintenance burdens in this module.

**Comment by bashtage** [comment]
In the case of dummy_buf it is whether the code path is correct for any functions that defer evaluation outside of the apply, which can result in incorrect results. This seems like a bigger issue.

**Comment by WillAyd** [comment]
Do you have a code sample to reproduce that issue? Can see if #34080 fixes it once green

**Comment by bashtage** [comment]
```
import statsmodels.api as sm
import pandas as pd
import numpy as np

np.random.seed(42)

Ys = pd.DataFrame(np.random.randn(4,10))
regressors = np.random.randn(10,2)

def regress1(s):
    return sm.OLS(s.values, regressors).fit()

results = pd.DataFrame(Ys).apply(regress1, axis='columns')
print([res.rsquared for res in results])

correct = [ sm.OLS(Ys.T[col], regressors).fit().rsquared for col in Ys.T]
print(correct)
```

Return
```
[-inf, -inf, -inf, -inf]
[0.4362196642459433, 0.011922654209731376, 0.31130804443466864, 0.16468214939479164]
c:\git\statsmodels\statsmodels\regression\linear_model.py:1702: RuntimeWarning: divide by zero encountered in double_scalars
  return 1 - self.ssr/self.uncentered_tss
```

The first return may vary since the data seems to be from an empty array which has random values.

**Comment by bashtage** [comment]
Not that simple, but it was how this issue was discovered.

**Comment by TomAugspurger** [comment]
What's the status here? We have a few places to update (can we compile a list in the original post? I'll start it) that we need to update before we can build pandas with Cython 3.x?

Is there anything that has to be done here for pandas 1.1.0, other than perhaps setting an upper bound for Cython in our pyproject.toml?

**Comment by WillAyd** [comment]
I think we need to invest some time to get things to work with Cython 3. Specifically this comment https://github.com/pandas-dev/pandas/issues/34014#issuecomment-624401636 I started down that path in #34080 just never saw through

**Comment by jbrockmendel** [comment]
> Yah, we need to decide how big a perf penalty we're willing to accept in exchange for losing the maintenance burdens in this module.

Re-upping this.  Just ripping out `Reducer` entirely and running asvs with `-b frame_apply` im getting

```
       before           after         ratio
     [72aed3e3]       [4b1dc5b1]
     <master>         <ref-no-reducer>
+     8.53±0.09ms       59.2±0.6ms     6.93  frame_methods.Apply.time_apply_ref_by_name
+         313±3ms          438±2ms     1.40  frame_methods.Nunique.time_frame_nunique
+         130±3ms          172±1ms     1.32  frame_methods.Apply.time_apply_user_func
+      11.2±0.2ms       14.1±0.3ms     1.25  frame_methods.Apply.time_apply_np_mean
+      10.5±0.1ms      12.8±0.08ms     1.23  frame_methods.Apply.time_apply_lambda_mean
-     30.1±0.08ms       26.2±0.1ms     0.87  frame_methods.Dropna.time_dropna_axis_mixed_dtypes('any', 1)
-     6.86±0.04ms      5.13±0.03ms     0.75  frame_methods.Apply.time_apply_pass_thru
```

**Comment by bashtage** [comment]
Can't you use `PyArray_New` to create a zero-copy array from an existing array without reassigning the data pointer?

**Comment by jbrockmendel** [comment]
@mroeschke any idea for ways to avoid the problem-causing pattern in aggregations.pyx?

**Comment by mroeschke** [comment]
Not immediately. It looks to affect `rolling.apply` but I can try to play around with it

**Comment by jbrockmendel** [comment]
AFAICT [this](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/reduction.pyx#L465) code

```
    cdef move(self, int start, int end):
        """
        For slicing
        """
        self.buf.data = self.values.data + self.stride * start
        self.buf.shape[0] = end - start
```

should be equivalent (ex the creation of a new ndarray object) to

```
    cdef move(self, int start, int end):
        self.buf = self.values[start:end]
```

but when I try this I get a bunch of test failures and a segfault that i cant reproduce in isolation.  Am I misunderstanding what `move` is doing?

**Comment by jbrockmendel** [comment]
Is this a blocker for 1.1?

**Comment by WillAyd** [comment]
I don’t think it needs to block a release as it’s a development dependency, though obviously one we want to keep current

**Comment by TomAugspurger** [comment]
I think if we don't update this for 1.1, then we'll want to set an upper bound on Cython in our pyproject.toml so that users can compile pandas from source.

**Comment by bashtage** [comment]
To advocate for fixing sooner rather than later, in addition to being a technical issue for future Cython, it is also a bug when the function being applied has any deferred operations.

**Comment by jbrockmendel** [comment]
@bashtage attempts so far have not gone great, xref #34997, extra eyeballs may help out

**Comment by fangchenli** [comment]
Here is one example that does not set data pointer in the `move` method.

```python
cdef class Slider:
    cdef:
        ndarray values, buf
        Py_ssize_t stride
        char *orig_data

    def __init__(self, ndarray values, ndarray buf):
        assert values.ndim == 1
        assert values.dtype == buf.dtype

        if not values.flags.contiguous:
            values = values.copy()

        self.values = values.copy()  # the original values would be modified if not copied
        self.buf = buf

        self.stride = values.strides[0]
        self.orig_data = self.buf.data

        self.buf.data = self.values.data
        self.buf.strides[0] = self.stride

    cdef move(self, int start, int end):
        # self.buf.data = self.values.data + self.stride * start
        self.buf.shape[0] = end - start  # length must be set first
        self.buf[:] = self.values[start:end]

    cdef reset(self):
        self.buf.data = self.orig_data
        self.buf.shape[0] = 0
```

The two pointer setting statements in the `__init__` and `reset` cannot be removed. Otherwise, there would be segfaults and free wrong memory errors. Somehow, we have to keep the original buf. It won't work if we do something like `self.buf = buf.copy()` or `self.buf = self.values[start:end]`. 

Unfortunately I don't know enough about cython and other parts of the code to figure this out. We might need to look at how the Slider's buf is used to find another way for this.

**Comment by tacaswell** [comment]
I spent some more time looking at what this is doing today and came up with the following notes (I apologize if I am saying really obvious things)
 
The `Slider` and `BlockSlider` classes are implementing views into a numpy array by:
 - taking in two arrays
 - stashing the pointer to one of them
 - on demand mutating the second array to point at the (offset) pointer to the first and adjusting the strides to give the window
 - on the way out the old guts of the second array are put back

The mutated array is then used to update "cached objects" in their calling class by updating the pandas side block manager details.  I suspect that this is the source of the stats model issues mentioned above as the code is aggressively changing things underneath the eventual user-exposed objects.

The change that has broken things is than cython now disallows relpacing the guts of a numpy array (which seem fair!).  My guess is that the performance gains come from both not memory thrashing and not falling back to the python layer.  The cython docs says that when you do `[]` on a numpy array it falls back to python (I assume because the inputs are too variable) which is probably the source of the major performance regressions.

I am not super clear how the numpy nbiter interface works, but it looks like it is focused on getting an iterator over single elements, or at least fixed steps through the array, where as for this code we need iteration over variable size windows.

It looks like the way to do this with memory views (  https://cython.readthedocs.io/en/latest/src/userguide/numpy_tutorial.html#efficient-indexing-with-memoryviews ) but those seem to require knowing what the type is up front. 

My suspicion is that the right solution here is to do something like what @mattip suggested above and in `def move` use the pointers we have to the underlying data and fabricate new numpy arrays of just the sub-section that is needed.

These classes appear to only be used internally to the reduction module so I do not think there are any back-compatibility with completely re-writing them.

attn @scoder for guidance on which of these methods (or one I do not see) is the best path.

**Comment by tacaswell** [comment]
also, I think this should be milestoned to an actual release.

**Comment by jreback** [comment]
> also, I think this should be milestoned to an actual release.

w/o a dedicated resource to do this it's impossible to actually say it will be in a particular release

**Comment by jbrockmendel** [comment]
> also, I think this should be milestoned to an actual release.

This would certainly be nice.  

Is there any indication that cython will be releasing 3.0 anytime soon?  The mailing list has been quiet; i havent been following the issue tracker recently.

I still think #34997 was a reasonable approach, but i never got it working.  More eyeballs may be helpful.

IIRC one of the hangups was that the non-cython path behaves slightly differently from the cython path when it comes to unwrapping results at each step in the loop.  If that were resolved, then we would at least have the option of ripping out the cython version altogether.

**Comment by fangchenli** [comment]
closed #43505

**Comment by tacaswell** [comment]
Well, "rip it all out" is one way to fix it !
